generator/src/main/java/com/graphicsfuzz/generator/tool/GenerateShaderFamily.java

```diff
@@ -0,0 +1,363 @@
+
+    Namespace ns = parse(args);
+
+    final String glslVersion = ns.getString("glsl_version");
+    final File referenceShaderJob = ns.get("reference_shader_job");
+    final File donorsDir = ns.get("donors");
+    final File outputDir = ns.get("output_dir") == null ? new File(".") : ns.get("output_dir");
+    final boolean webgl = ns.getBoolean("webgl");
+    final boolean verbose = ns.getBoolean("verbose");
+    final boolean disableValidator = ns.getBoolean("disable_validator");
+    final boolean writeProbabilities = ns.getBoolean("write_probabilities");
+    final boolean keepBadVariants = ns.getBoolean("keep_bad_variants");
+    final boolean stopOnFail = ns.getBoolean("stop_on_fail");
+    final int seed = ns.getInt("seed");
+    final int numVariants = ns.getInt("num_variants");
+    Optional<Integer> maxBytes = ns.get("max_bytes") == null ? Optional.empty() :
+        Optional.of(ns.getInt("max_bytes"));
+    Optional<Float> maxFactor = ns.get("max_factor") == null ? Optional.empty() :
+        Optional.of(ns.getFloat("max_factor"));
+    final GeneratorArguments generatorArguments = Generate.getGeneratorArguments(ns);
+
+    if (webgl && !ShadingLanguageVersion.isWebGlCompatible(glslVersion)) {
+      // Check immediately that if --webgl has been passed, it is possible to get the given
+      // WebGL shading language version as a string.
+      throw new RuntimeException("Given GLSL version " + glslVersion + " is not WebGL-compatible.");
+    }
+
+    if (verbose) {
+      LOGGER.info("Using seed " + seed);
+    }
+
+    if (!referenceShaderJob.isFile()) {
+      throw new FileNotFoundException("Reference shader job " + referenceShaderJob.getAbsolutePath()
+          + " does not exist.");
+    }
+
+    if (verbose) {
+      LOGGER.info("Using donor folder " + donorsDir.getAbsolutePath());
+    }
+
+    FileUtils.forceMkdir(outputDir);
+
+    final ShaderJobFileOperations fileOps = new ShaderJobFileOperations();
+
+    // Prepare reference shaders.
+    final File preparedReferenceShaderJob = new File(outputDir, "reference.json");
+    PrepareReference.prepareReference(referenceShaderJob,
+        preparedReferenceShaderJob,
+        ShadingLanguageVersion.fromVersionString(glslVersion),
+        generatorArguments.getReplaceFloatLiterals(),
+        // We subtract 1 because we need to be able to add injectionSwitch
+        generatorArguments.getMaxUniforms() - 1,
+        generatorArguments.getGenerateUniformBindings(),
+        fileOps);
+
+    // Validate reference shaders.
+    if (!disableValidator) {
+      if (!(fileOps.areShadersValid(preparedReferenceShaderJob, false)
+          && fileOps.areShadersValidShaderTranslator(preparedReferenceShaderJob, false))) {
+        throw new RuntimeException("One or more of the prepared shaders of shader job "
+            + preparedReferenceShaderJob.getAbsolutePath() + " is not valid.");
+      }
+    }
+
+    if (primitivesFile(referenceShaderJob).isFile()) {
```

Comment:
Maybe can use: ShaderJobFileOperations.copyShaderJobFileTo.
And just generally: no direct file operations if possible.

Reply:
Leaving as is for now, since primitives is a bit complex and we might get rid of it anyway.